pkg/monitor/filemonitor.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package monitor
+
+import (
+	"os"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/rs/zerolog"
+)
+
+type Subscription struct {
+	Entity       Subscriber
+	Options      FileMonitorOptions
+	LastModified time.Time
+	NumFiles     int
+}
+
+// FileMonitor detects  changes since the last check and notifies subscribers
+type FileMonitor struct {
+	// log
+	Log zerolog.Logger
+	// list of subscribers and their states
+	Subsciptions []Subscription
+}
+
+type FileMonitorOptions struct {
+	// Directory that should be monitored for change
+	Path string
+	// File extension
+	Extension string
+}
+
+// Subscribe to get notifications on file changes
+func (m *FileMonitor) Subscribe(subscriber Subscriber) error {
+	s := Subscription{
+		Entity:  subscriber,
+		Options: subscriber.GetOptions(),
+	}
+	numFiles, modified, err := m.visit(&s)
+	if err != nil {
+		return err
+	}
+	s.LastModified = modified
+	s.NumFiles = numFiles
+	m.Subsciptions = append(m.Subsciptions, s)
+	return nil
+}
+
+// Run activates a go routine that calls Monitor() upon changes in /tmp/adminconfig directory
+func (m *FileMonitor) Run(watcher *fsnotify.Watcher) {
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					continue
+				}
+				m.Log.Info().Msg("Event: " + event.String())
+				m.Monitor()
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					continue
+				}
+				m.Log.Err(err).Msg("error watching file changes")
+			}
+		}
+	}()
+}
+
+// Check if changes in files require any action
+func (m *FileMonitor) Monitor() {
```

Comment:
why do we scan all files and validate their timestamps?  we can get the same info from the `watcher.Event.Name` and `watcher.Event.Op` . 
```go
// Event represents a single file system notification.
type Event struct {
	Name string // Relative path to the file or directory.
	Op   Op     // File operation that triggered the event.
}

// Op describes a set of file operations.
type Op uint32

// These are the generalized file operations that can trigger a notification.
const (
	Create Op = 1 << iota
	Write
	Remove
	Rename
	Chmod
)
```

Reply:
It won't work since the directory is not watched recursively.
This is the content of the watched directory:
<timestamp>
..data -> <timestamp>
*. json/rego -> ../data/*.json/rego

We are getting events about creating <new_timestamp>, deleting <old_timestamp> and modifying ..data folders. That's all. The actual files are inside <timestamp> directory and we won't know what has changed.

Reply:
you're right that fsnotify doesn't support recursive watcher (it has been in the roadmap for a long time). 
However:

- if you get an event about directory creation (we can verify that a new file is a directory) you can add this directory to the root watcher, or create a new one.
- all these directories timestamps modification is done by the manager, co it can be changed or it can provide input to the monitor

You have implemented the monitor, but my question is - given fsnotify, can we simplify the entire process and prevent scanning all files?

Reply:
Another comment, like you said, for each "update" we get several events, like create/delete <timestamp> and write in data folders and for each of them the `Monitor` method is run. 
Suggestion, use a predefined filename prefix for "timestamps dir name", and run `Monitor` only on 'REMOVE` events of files with this prefix.

Reply:
I am using fsnotify to avoid polling for changes. However, since all files are in the same directory, and we don't need to compile rego when json fles change, I still need to understand where the change was. 
I don't control the timestamps dir, it is created automatically when a configmap change is detected and the pod gets an update. 
The sequence of the events when running as a pod is different from running locally (for tests or debug) and depends on the internal implementation of the fsnotify. The more general approach is to react to all events. Also, I don't think that reading timestamps of a few files is that much of an overhead.

Reply:
Ignoring CHMOD event (others are required)
Added an integration test that demonstrates change in the policies (a new policy file is added) and how it affects the data plane.